File: frontend/src/types/changelog.ts
/**
 * Changelog types for Story 9.7 - In-App Changelog & Release Notes
 *
 * Provides type definitions for changelog entries and read status tracking.
 */

// =============================================================================
// Enum Types
// =============================================================================

/**
 * Types of changelog releases.
 */
export type ReleaseType = 'feature' | 'improvement' | 'fix' | 'deprecation' | 'security';

/**
 * Feature areas for contextual badge targeting.
 */
export type FeatureArea =
  | 'dashboard'
  | 'sync_health'
  | 'insights'
  | 'recommendations'
  | 'approvals'
  | 'connectors'
  | 'billing'
  | 'settings'
  | 'reports'
  | 'notifications';

// =============================================================================
// Data Interfaces
// =============================================================================

/**
 * A changelog entry as returned by the API.
 */
export interface ChangelogEntry {
  id: string;
  version: string;
  title: string;
  summary: string;
  content?: string;
  release_type: ReleaseType;
  feature_areas: FeatureArea[];
  published_at?: string;
  documentation_url?: string;
  is_read: boolean;
}

/**
 * Admin view of a changelog entry with additional metadata.
 */
export interface ChangelogAdminEntry extends ChangelogEntry {
  is_published: boolean;
  created_by_user_id: string;
  created_at: string;
  updated_at: string;
}

// =============================================================================
// Response Interfaces
// =============================================================================

/**
 * Response for changelog list queries.
 */
export interface ChangelogListResponse {
  entries: ChangelogEntry[];
  total: number;
  has_more: boolean;
  unread_count: number;
}

/**
 * Response for admin changelog list queries.
 */
export interface ChangelogAdminListResponse {
  entries: ChangelogAdminEntry[];
  total: number;
  has_more: boolean;
}

/**
 * Response for unread count query.
 */
export interface ChangelogUnreadCountResponse {
  count: number;
  by_feature_area: Record<string, number>;
}

/**
 * Response after marking entries as read.
 */
export interface ChangelogMarkReadResponse {
  marked_count: number;
  unread_count: number;
}

// =============================================================================
// Request/Filter Interfaces
// =============================================================================

/**
 * Filters for changelog list queries.
 */
export interface ChangelogFilters {
  release_type?: ReleaseType;
  feature_area?: FeatureArea;
  include_read?: boolean;
  limit?: number;
  offset?: number;
}

/**
 * Request to create a new changelog entry (admin only).
 */
export interface ChangelogCreateRequest {
  version: string;
  title: string;
  summary: string;
  content?: string;
  release_type: ReleaseType;
  feature_areas?: FeatureArea[];
  documentation_url?: string;
}

/**
 * Request to update a changelog entry (admin only).
 */
export interface ChangelogUpdateRequest {
  version?: string;
  title?: string;
  summary?: string;
  content?: string;
  release_type?: ReleaseType;
  feature_areas?: FeatureArea[];
  documentation_url?: string;
}

// =============================================================================
// Utility Functions
// =============================================================================

/**
 * Get human-readable label for a release type.
 */
export function getReleaseTypeLabel(type: ReleaseType): string {
  const labels: Record<ReleaseType, string> = {
    feature: 'New Feature',
    improvement: 'Improvement',
    fix: 'Bug Fix',
    deprecation: 'Deprecation',
    security: 'Security',
  };
  return labels[type] || type;
}

/**
 * Get badge tone for a release type (for Polaris Badge).
 */
export function getReleaseTypeTone(
  type: ReleaseType
): 'info' | 'success' | 'warning' | 'critical' | 'attention' | undefined {
  const tones: Record<ReleaseType, 'info' | 'success' | 'warning' | 'critical' | 'attention' | undefined> = {
    feature: 'success',
    improvement: 'info',
    fix: 'attention',
    deprecation: 'warning',
    security: 'critical',
  };
  return tones[type];
}

/**
 * Get icon name for a release type.
 */
export function getReleaseTypeIcon(type: ReleaseType): string {
  const icons: Record<ReleaseType, string> = {
    feature: 'StarFilledIcon',
    improvement: 'ArrowUpIcon',
    fix: 'WrenchIcon',
    deprecation: 'AlertCircleIcon',
    security: 'LockIcon',
  };
  return icons[type] || 'InfoIcon';
}

/**
 * Get human-readable label for a feature area.
 */
export function getFeatureAreaLabel(area: FeatureArea): string {
  const labels: Record<FeatureArea, string> = {
    dashboard: 'Dashboard',
    sync_health: 'Sync Health',
    insights: 'AI Insights',
    recommendations: 'Recommendations',
    approvals: 'Approvals',
    connectors: 'Connectors',
    billing: 'Billing',
    settings: 'Settings',
    reports: 'Reports',
    notifications: 'Notifications',
  };
  return labels[area] || area;
}

/**
 * Format a date string for display.
 */
export function formatChangelogDate(dateString?: string): string {
  if (!dateString) return '';
  const date = new Date(dateString);
  return date.toLocaleDateString('en-US', {
    year: 'numeric',
    month: 'short',
    day: 'numeric',
  });
}

/**
 * Group changelog entries by version.
 */
export function groupEntriesByVersion(entries: ChangelogEntry[]): Map<string, ChangelogEntry[]> {
  const grouped = new Map<string, ChangelogEntry[]>();
  for (const entry of entries) {
    const existing = grouped.get(entry.version) || [];
    existing.push(entry);
    grouped.set(entry.version, existing);
  }
  return grouped;
}

File: backend/src/api/schemas/changelog.py
"""
Changelog API schemas for Story 9.7.

Pydantic models for changelog entries and read status.

SECURITY:
- Admin-only write operations
- Any authenticated user can read
"""

from datetime import datetime
from typing import Optional, List

from pydantic import BaseModel, Field, field_validator


class ChangelogEntryResponse(BaseModel):
    """Single changelog entry for display."""

    id: str
    version: str
    title: str
    summary: str
    content: Optional[str] = None
    release_type: str
    feature_areas: List[str] = Field(default_factory=list)
    published_at: Optional[datetime] = None
    documentation_url: Optional[str] = None
    is_read: bool = False  # Populated per-user


class ChangelogListResponse(BaseModel):
    """Response for changelog list queries."""

    entries: List[ChangelogEntryResponse]
    total: int
    has_more: bool
    unread_count: int


class ChangelogUnreadCountResponse(BaseModel):
    """Response for unread count query."""

    count: int
    by_feature_area: dict[str, int] = Field(default_factory=dict)


class ChangelogCreateRequest(BaseModel):
    """Request to create a new changelog entry (admin only)."""

    version: str = Field(..., min_length=1, max_length=50)
    title: str = Field(..., min_length=1, max_length=500)
    summary: str = Field(..., min_length=1)
    content: Optional[str] = None
    release_type: str = Field(..., pattern=r"^(feature|improvement|fix|deprecation|security)$")
    feature_areas: List[str] = Field(default_factory=list)
    documentation_url: Optional[str] = Field(None, max_length=1000)

    @field_validator("feature_areas")
    @classmethod
    def validate_feature_areas(cls, v: List[str]) -> List[str]:
        """Validate feature areas are from allowed list."""
        from src.models.changelog_entry import FEATURE_AREAS
        invalid = set(v) - set(FEATURE_AREAS)
        if invalid:
            raise ValueError(f"Invalid feature areas: {invalid}. Allowed: {FEATURE_AREAS}")
        return v


class ChangelogUpdateRequest(BaseModel):
    """Request to update a changelog entry (admin only)."""

    version: Optional[str] = Field(None, min_length=1, max_length=50)
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    summary: Optional[str] = Field(None, min_length=1)
    content: Optional[str] = None
    release_type: Optional[str] = Field(None, pattern=r"^(feature|improvement|fix|deprecation|security)$")
    feature_areas: Optional[List[str]] = None
    documentation_url: Optional[str] = Field(None, max_length=1000)

    @field_validator("feature_areas")
    @classmethod
    def validate_feature_areas(cls, v: Optional[List[str]]) -> Optional[List[str]]:
        """Validate feature areas are from allowed list."""
        if v is None:
            return v
        from src.models.changelog_entry import FEATURE_AREAS
        invalid = set(v) - set(FEATURE_AREAS)
        if invalid:
            raise ValueError(f"Invalid feature areas: {invalid}. Allowed: {FEATURE_AREAS}")
        return v


class ChangelogAdminEntryResponse(BaseModel):
    """
    Changelog entry for admin view.

    Includes additional fields not shown to regular users.
    """

    id: str
    version: str
    title: str
    summary: str
    content: Optional[str] = None
    release_type: str
    feature_areas: List[str] = Field(default_factory=list)
    is_published: bool
    published_at: Optional[datetime] = None
    documentation_url: Optional[str] = None
    created_by_user_id: str
    created_at: datetime
    updated_at: datetime


class ChangelogAdminListResponse(BaseModel):
    """Response for admin changelog list queries."""

    entries: List[ChangelogAdminEntryResponse]
    total: int
    has_more: bool


class ChangelogMarkReadRequest(BaseModel):
    """Request to mark entries as read."""

    entry_ids: List[str] = Field(default_factory=list)


class ChangelogMarkReadResponse(BaseModel):
    """Response after marking entries as read."""

    marked_count: int
    unread_count: int
